### PR TITLE
Target List request performance optimisation

### PR DESCRIPTION
**Pull request by hugoamvieira**
This PR is the `Targets` counterpart of https://github.com/hashicorp/boundary/pull/2381 and contains a set of commits designed to improve Boundary's performance when listing Target resources.

### CODE REVIEW

File: internal/perms/acl.go [hunk 1]
```diff
@@ -209,6 +219,72 @@ func (a ACL) Allowed(r Resource, aType action.Type, userId string, opt ...Option
...
 }
 
+// AllowedListConditions builds a set of user permissions (according to their grants)
+// for resource listing ahead of DB access.
+// This enables list query optimization: instead of having to query all data and then drop
+// what the requestor is not authorized to see, we can use the data in the Permission objects
+// to build a more efficient DB query.
+// It does not (and is not meant to) override current authorization mechanisms.
+//
+// TBD: New name - BuildListingPermissions ?
```
Comment (hugoamvieira) [inline internal/perms/acl.go:229]: TBD

File: internal/perms/acl.go [hunk 2]
```diff
@@ -27,6 +28,15 @@ type ACLResults struct {
 	scopeMap map[string][]Grant
 }
 
+// Permission objects are built off of a set of Boundary grants, and serve to
+// enable database query optimization, specifically for Boundary resource listing.
+type Permission struct {
+	ScopeId  string   // The scope id for which the permission applies.
+	Ids      []string // Any specific resource ids that have been referred in the grant's `id` field, if applicable.
+	OnlySelf bool     // The grant only allows actions against the user's own resources. TBD: This is only used for list and there's no `list:self`.
```
Comment (hugoamvieira) [inline internal/perms/acl.go:36]: TBD

Comment (hugoamvieira) [inline internal/perms/acl.go:36]: TBD

File: internal/perms/acl.go (hunk 1 above)
Comment (hugoamvieira) [inline internal/perms/acl.go:229]: TBD

File: internal/perms/acl.go [hunk 3]
```diff
@@ -209,6 +219,72 @@ func (a ACL) Allowed(r Resource, aType action.Type, userId string, opt ...Option
...
+						found = true
+						break
+					}
+				}
+			}
+			if !found { // In this case, no actions were granted for the given scope id.
+				continue
+			}
+
+			// TBD: `list:self` isn't a thing. Considering we're listing here, do we need this?
```
Comment (hugoamvieira) [inline internal/perms/acl.go:260]: TBD

File: internal/target/options.go [hunk 4]
```diff
@@ -178,3 +181,11 @@ func WithTargetIds(with []string) Option {
 		o.WithTargetIds = with
 	}
 }
+
+// WithListPermissions enables this repo to optimise resource listing
+// database queries.
```
Comment (tmessi) [inline internal/target/options.go:186]: While this does help with optimizing the resource listing, I think it primarily is used to restrict the list results based on this set of permissions.

File: internal/target/repository.go [hunk 5]
```diff
@@ -188,66 +198,31 @@ func (r *Repository) FetchAuthzProtectedEntitiesByScope(ctx context.Context, sco
...
-			idsInClause, args = append(idsInClause, fmt.Sprintf("@%d", inClauseCnt)), append(args, sql.Named(fmt.Sprintf("%d", inClauseCnt), id))
-		}
-		where = append(where, fmt.Sprintf("public_id in (%s)", strings.Join(idsInClause, ",")))
-	}
-
-	if opts.WithType != "" {
-		inClauseCnt += 1
-		where, args = append(where, fmt.Sprintf("type = @%d", inClauseCnt)), append(args, sql.Named(fmt.Sprintf("%d", inClauseCnt), opts.WithType.String()))
+	if len(r.listPermissions) == 0 {
+		return nil, errors.New(ctx, errors.Internal, op, "failed to list targets")
```
Comment (tmessi) [inline internal/target/repository.go:206]: I am not sure we should error if `r.listPermissions` is empty. I think returning an empty list might make sense.

File: internal/target/repository.go [hunk 6]
```diff
@@ -269,6 +244,32 @@ func (r *Repository) list(ctx context.Context, resources interface{}, where stri
 	return nil
 }
 
+func (r *Repository) permissionWhereClauses() ([]string, []interface{}) {
```
Comment (tmessi) [inline internal/target/repository.go:247]: Since this is specific to list permissions, you might want to name this method `listPermissionWhereClauses`

**Review by tmessi — CHANGES_REQUESTED** [review]
couple minor things, otherwise looking good

File: internal/target/options.go (hunk 4 above)
Comment (hugoamvieira) [reply internal/target/options.go]: Yep, thanks, addressed in [aa56f90](https://github.com/hashicorp/boundary/pull/2319/commits/aa56f905816e7b4d6b1eb9e04ea5ba85fc61cb20)

File: internal/target/repository.go (hunk 6 above)
Comment (hugoamvieira) [reply internal/target/repository.go]: Addressed in [aa56f90](https://github.com/hashicorp/boundary/pull/2319/commits/aa56f905816e7b4d6b1eb9e04ea5ba85fc61cb20)

File: internal/target/repository.go (hunk 5 above)
Comment (hugoamvieira) [reply internal/target/repository.go]: Agreed - Addressed in [0012e01](https://github.com/hashicorp/boundary/pull/2319/commits/0012e01dfe33497ceebd49eb7470b2399722ebc2)

File: internal/target/repository.go [hunk 7]
```diff
@@ -188,66 +198,31 @@ func (r *Repository) FetchAuthzProtectedEntitiesByScope(ctx context.Context, sco
 	return targetsMap, nil
 }
 
-// ListTargets in targets in a scope.  Supports the WithScopeId, WithLimit, WithType options.
+// ListTargets in targets in a scope.
```
Comment (tmessi) [inline internal/target/repository.go:201]: Can you update this go doc? What is currently here does not make sense to me. It should also note some about the impact of `listPermissions` on the results.

**Review by tmessi — CHANGES_REQUESTED** [review]
some minor doc things, otherwise I think this is good to get into the llb.

File: internal/target/repository.go [hunk 8]
```diff
@@ -49,10 +58,11 @@ func NewRepository(r db.Reader, w db.Writer, kms *kms.Kms, opt ...Option) (*Repo
...
 	return &Repository{
-		reader:       r,
-		writer:       w,
-		kms:          kms,
-		defaultLimit: opts.WithLimit,
+		reader:          r,
+		writer:          w,
+		kms:             kms,
+		defaultLimit:    opts.WithLimit,
+		listPermissions: opts.WithListPermissions,
```
Comment (tmessi) [inline internal/target/repository.go:65]: I think this should check that all of the permissions in `opts.WithListPemrissions` have `action.List` and `resource.Target`.

File: internal/daemon/controller/auth/auth.go [hunk 9]
```diff
@@ -696,6 +696,10 @@ func GetTokenFromRequest(ctx context.Context, kmsCache *kms.Kms, req *http.Reque
 	return publicId, encryptedToken, uint32(receivedTokenType)
 }
 
+func (r *VerifyResults) ACL() perms.ACL {
+	return r.v.acl
```
[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/daemon/controller/auth/auth.go:705]: Sorry about all the git wrangling - Here's the change [5ee3b51](https://github.com/hashicorp/boundary/pull/2319/commits/5ee3b51a16230e0bacfe136a24e7490a6328d135)

File: internal/daemon/controller/auth/auth.go [hunk 10]
```diff
@@ -696,6 +696,10 @@ func GetTokenFromRequest(ctx context.Context, kmsCache *kms.Kms, req *http.Reque
 	return publicId, encryptedToken, uint32(receivedTokenType)
 }
 
+func (r *VerifyResults) ACL() perms.ACL {
```
[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/daemon/controller/auth/auth.go:700]: Added in [5ee3b51](https://github.com/hashicorp/boundary/pull/2319/commits/5ee3b51a16230e0bacfe136a24e7490a6328d135)

File: internal/target/repository.go [hunk 11]
```diff
@@ -12,9 +12,12 @@ import (
 	"github.com/hashicorp/boundary/internal/errors"
 	"github.com/hashicorp/boundary/internal/kms"
 	"github.com/hashicorp/boundary/internal/oplog"
+	"github.com/hashicorp/boundary/internal/perms"
 	"github.com/hashicorp/boundary/internal/types/scope"
 )
 
+type RepositoryFactory func(...Option) (*Repository, error)
```
[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/target/repository.go:23]: Added in [7709999](https://github.com/hashicorp/boundary/pull/2319/commits/7709999475ddbce1bad1cf98d0833d0d612d9ca9)

File: internal/target/repository.go (hunk 7 above)
Comment (hugoamvieira) [reply internal/target/repository.go]: Yep - Good point, added in https://github.com/hashicorp/boundary/pull/2319/commits/77810b9b59d88072eddef041e99a459218f4880a#diff-68b2bab8e324477709a446d45241a2dcf18c53519dcdad0718f97d0c30f6c161R216-R219

**Review by mgaffney — APPROVED** [review]

**Review by tmessi — APPROVED** [review]

**Review by tmessi — APPROVED** [review]

**Comment by jefferai** [comment]
This is the analogue to Tim's session listing PR right?

**Comment by tmessi** [comment]
> This is the analogue to Tim's session listing PR right?

Yep

**Comment by jefferai** [comment]
Sounds good, I just started reviewing that and will check this one out after!

**Comment by hugoamvieira** [comment]
Merged into https://github.com/hashicorp/boundary/pull/2342

Any reviewers that would like to review this and weren't able to do so, please take a look at the LLB.

**Pull request merged**